src/server/src/main/java/io/cassandrareaper/service/SegmentGenerator.java

```diff
@@ -81,76 +87,186 @@ static boolean greaterThanOrEqual(BigInteger big0, BigInteger big1) {
-      BigInteger[] segmentCountAndRemainder
-          = rs.multiply(BigInteger.valueOf(totalSegmentCount)).divideAndRemainder(rangeSize);
+        int segmentCount =
+            segmentCountAndRemainder[0].intValue()
+                + (segmentCountAndRemainder[1].equals(BigInteger.ZERO) ? 0 : 1);
 
-      int segmentCount = segmentCountAndRemainder[0].intValue()
-          + (segmentCountAndRemainder[1].equals(BigInteger.ZERO) ? 0 : 1);
+        LOG.info("Dividing token range [{},{}) into {} segments", start, stop, segmentCount);
 
-      LOG.info("Dividing token range [{},{}) into {} segments", start, stop, segmentCount);
+        // Make big0 list of all the endpoints for the repair segments, including both start and stop
+        List<BigInteger> endpointTokens = Lists.newArrayList();
+        for (int j = 0; j <= segmentCount; j++) {
+          BigInteger offset =
+              rs.multiply(BigInteger.valueOf(j)).divide(BigInteger.valueOf(segmentCount));
+          BigInteger reaperToken = start.add(offset);
+          if (greaterThan(reaperToken, rangeMax)) {
+            reaperToken = reaperToken.subtract(rangeSize);
+          }
+          endpointTokens.add(reaperToken);
+        }
 
-      // Make big0 list of all the endpoints for the repair segments, including both start and stop
-      List<BigInteger> endpointTokens = Lists.newArrayList();
-      for (int j = 0; j <= segmentCount; j++) {
-        BigInteger offset = rs.multiply(BigInteger.valueOf(j)).divide(BigInteger.valueOf(segmentCount));
-        BigInteger reaperToken = start.add(offset);
-        if (greaterThan(reaperToken, rangeMax)) {
-          reaperToken = reaperToken.subtract(rangeSize);
+        // Append the segments between the endpoints
+        for (int j = 0; j < segmentCount; j++) {
+          repairSegments.add(
+              Segment.builder()
+                  .withTokenRanges(
+                      Arrays.asList(
+                          new RingRange(endpointTokens.get(j), endpointTokens.get(j + 1))))
+                  .build());
+          LOG.debug(
+              "Segment #{}: [{},{})", j + 1, endpointTokens.get(j), endpointTokens.get(j + 1));
         }
-        endpointTokens.add(reaperToken);
       }
 
-      // Append the segments between the endpoints
-      for (int j = 0; j < segmentCount; j++) {
-        repairSegments.add(new RingRange(endpointTokens.get(j), endpointTokens.get(j + 1)));
-        LOG.debug("Segment #{}: [{},{})", j + 1, endpointTokens.get(j), endpointTokens.get(j + 1));
+      // verify that the whole range is repaired
+      BigInteger total = BigInteger.ZERO;
+      for (Segment segment : repairSegments) {
+        for (RingRange range : segment.getTokenRanges()) {
+          BigInteger size = range.span(rangeSize);
+          total = total.add(size);
+        }
       }
+      if (!total.equals(rangeSize) && !incrementalRepair) {
+        throw new ReaperException("Not entire ring would get repaired");
```

Comment:
is this really a checked exception? how is the caller expected to deal with it and resume/repeat? how could this happen? maybe it's a IllegalStateException or an AssertionError?

Reply:
This is out of the scope of this PR, as this code was already there before.